sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -609,6 +611,16 @@ case class FileSourceScanExec(
     override val disableBucketedScan: Boolean = false)
   extends FileSourceScanLike {
 
+  /** SQL metrics generated for JSON scans involving variants. */
+  protected lazy val variantBuilderMetrics: Map[String, SQLMetric] =
+    VariantConstructionMetrics.createSQLMetrics(sparkContext)
+
+  /** Only report variant metrics if the data source file format is JSON */
+  override lazy val metrics: Map[String, SQLMetric] = super.metrics ++ {
+    if (relation.fileFormat.isInstanceOf[JsonFileFormat]) variantBuilderMetrics
```

Comment:
does json scan always produce variants? If not we should only display these metrics when variant will be produced by the scan.